src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java

```diff
@@ -0,0 +1,426 @@
+
+    this.connectorConfig = connectorConfig;
+    // Streaming ingest requires a properties.
+    // TODO request a builder pattern for SI Client.
+    Properties streamingClientProps = new Properties();
+    streamingClientProps.putAll(connectorConfig);
+    this.taskId = connectorConfig.getOrDefault(Utils.TASK_ID, "-1");
+    this.streamingIngestClientName =
+        STREAMING_CLIENT_PREFIX_NAME + conn.getConnectorName() + "_" + taskId;
+    this.streamingIngestClient =
+        SnowflakeStreamingIngestClientFactory.builder(this.streamingIngestClientName)
+            .setProperties(streamingClientProps)
+            .build();
+    this.partitionsToChannel = new HashMap<>();
+  }
+
+  @Override
+  public void startTask(String tableName, String topic, int partition) {
+    String partitionChannelKey = partitionChannelKey(topic, partition);
+    LOGGER.info(
+        "Opening a channel with name:{} for table name:{}, topic:{}, partition:{}",
+        partitionChannelKey,
+        tableName,
+        topic,
+        partition);
+    // the table should be present before opening a channel so lets do a table existence check here
+    createTableIfNotExists(tableName);
+    SnowflakeStreamingIngestChannel partitionChannel =
+        streamingIngestClient.openChannel(getOpenChannelRequest(partitionChannelKey, tableName));
+    TopicPartitionChannel topicPartitionChannel =
+        new TopicPartitionChannel(partitionChannel, this.conn, tableName);
+    partitionsToChannel.putIfAbsent(partitionChannelKey, topicPartitionChannel);
+  }
+
+  /**
+   * Inserts the given record into buffer and then eventually calls insertRows API if buffer
+   * threshold has reached.
+   *
+   * <p>TODO: SNOW-473896 - Please note we will get away with Buffering logic in future commits.
+   *
+   * @param records record content
+   */
+  @Override
+  public void insert(Collection<SinkRecord> records) {
+    // note that records can be empty
+    for (SinkRecord record : records) {
+      // check if need to handle null value records
+      if (recordService.shouldSkipNullValue(record, behaviorOnNullValues)) {
+        continue;
+      }
+      // Might happen a count of record based flushing
+      insert(record);
+    }
+    // check all sink context to see if they need to be flushed
+    for (TopicPartitionChannel partitionChannel : partitionsToChannel.values()) {
+      // Time based flushing
+      if ((System.currentTimeMillis() - partitionChannel.getPreviousFlushTimeStampMs())
+          >= (getFlushTime() * 1000)) {
+        LOGGER.info("Time based flush for channel:{}", partitionChannel.getChannelName());
+        LOGGER.info(
+            "Current:{}, previousFlushTime:{}, threshold:{}",
+            System.currentTimeMillis(),
+            partitionChannel.getPreviousFlushTimeStampMs(),
+            (getFlushTime() * 1000));
+        partitionChannel.insertBufferedRows();
+      }
+    }
```

Comment:
This is interesting, ideally we want a background thread to do this, I assume this is probably the old behavior?

Reply:
so lines 169-181
- It is time based flush.
- Every 5second or so we have been invoked with put API which checks if time based flush has been hit. 
- It is kind of unimportant right now since we will not have any flush. (i.e we will not have any buffer)

Also, yes you are right, this is the V1 behavior.

Reply:
hmmm, I thought the conclusion was that the first version will still honor the flush time/buffer size, meaning we will have two buffers?

Reply:
Update: We will keep the buffer params, essentially saying we will have double buffer. 
As far as BG thread is concerned, we dont necessarily need to do for PrPr, will log a JIRA and see if this can be done.